core/src/main/java/org/apache/iceberg/CherryPickFromSnapshot.java

```diff
@@ -0,0 +1,225 @@
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+/**
+ * In an audit workflow, new data is written to an orphan snapshot that is not committed as the table's
+ * current state until it is audited. After auditing a change, it may need to be applied or cherry-picked
+ * on top of the latest snapshot instead of the one that was current when the audited changes were created.
+ *
+ * This class adds support for cherry-picking the changes from an orphan snapshot by applying them to
+ * the current snapshot. The output of the operation is a new snapshot with the changes from cherry-picked
+ * snapshot.
+ *
+ * Cherry-picking should apply the exact set of changes that were done in the original commit.
+ *  - All added files should be added to the new version.
+ *  - Todo: If files were deleted, then those files must still exist in the data set.
+ *  - Does not support Overwrite operations currently. Overwrites are considered as conflicts.
+ *
+ */
+class CherryPickFromSnapshot extends MergingSnapshotProducer<AppendFiles> implements CherryPick {
+  private final TableOperations ops;
+  private TableMetadata base = null;
+  private Long cherryPickSnapshotId = null;
+  private final AtomicInteger manifestCount = new AtomicInteger(0);
+
+  CherryPickFromSnapshot(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.refresh();
+  }
+
+  @Override
+  protected AppendFiles self() {
+    throw new UnsupportedOperationException("Not supported");
+  }
+
+  /**
+   * We only cherry pick for appends right now
+   * @return
+   */
+  @Override
+  protected String operation() {
+    return DataOperations.APPEND;
+  }
+
+  @Override
+  public CherryPickFromSnapshot fromSnapshotId(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    this.cherryPickSnapshotId = snapshotId;
+    return this;
+  }
+
+  /**
+   * Apply the pending changes and return the uncommitted changes for validation.
+   * <p>
+   * This does not result in a permanent update.
+   *
+   * @return the uncommitted changes that would be committed by calling {@link #commit()}
+   * @throws ValidationException      If the pending changes cannot be applied to the current metadata
+   * @throws IllegalArgumentException If the pending changes are conflicting or invalid
+   */
+  @Override
+  public Snapshot apply() {
```

Comment:
Subclasses of `MergingSnapshotProducer` shouldn't override `apply` or `apply(TableMetadata)`. Those methods do the heavy lifting of producing and merging manifests, as well as handling manifest lists.

If you want to override `apply` then I would recommend doing that only to call `super.apply()` after configuring changes using the methods provided by `MergingSnapshotProducer`, like `add` and `delete`. The pattern most subclasses use is to call those methods as the caller adds configuration. For example, `MergeAppend` calls `add` when `appendFile` is called.

Reply:
Got it.. realised i was doing something funky there. So if i want to create a new snapshot with only appends from the cherrypick snapshot, i can just iterate over the `addedFiles` and call `add()` on those so that when i do `super.apply()` it would just get me a new snapshot with the appended files along with the previous snapshots' files/manifests merged in. 

Updating that now.

Reply:
Yeah, but if we want to use `addedFiles` then we should update it to project all columns. If you rewrote a manifest with a subset of DataFile columns then you'd lose some metadata.